Plugin Extensions: Only load app plugins when necessary

leventebalogh opened the pull request:
### What changed?

**This is not a breaking change.**

This PR removes the need to define `preload: true` for plugins that are using extensions. They will be automatically loaded in the background in case it's necessary:
- they are accessed directly (e.g. `/a/<PLUGIN_ID>/*`)
- they are extending an extension point that is being rendered (`usePluginLinks()` or `usePluginComponents()`)
- they are exposing a component that is being used (via `usePluginComponent()`)
- they are a dependency of a plugin (due to an exposed component) that is being loaded for one of the reasons above

<img width="1483" alt="Screenshot 2024-10-10 at 14 48 26" src="https://github.com/user-attachments/assets/edf6fd11-4865-441c-ba6a-4a95914a2e5b">
<img width="2555" alt="Screenshot 2024-11-05 at 13 57 29" src="https://github.com/user-attachments/assets/5b5ae3f3-a97a-4f81-9437-34e9e995379c">

diegommm commented:
Hi! Greetings from the Release Guild. Just letting you know I'm updating your PR's milestone to [11.2.x](https://github.com/grafana/grafana/milestone/529). If you need to see your changes in a previous version, please consider adding a backport label.

Thank you!

ifrost commented:
Hi @leventebalogh 👋  This item has been on our board for some time (Tracing and Profiling). Is it work in progress? Do you need any input from Tracing & Profiling?

leventebalogh committed:
feat(plugins): automatically preload plugins

This PR enables auto-preloading for plugins when they are used
by an extension or extension-point. Once this change is merged plugins
that were only using "preload: true" in their plugin.json for using extensions
can remove it.

torkelo commented on public/app/features/dashboard-scene/scene/PanelMenuBehavior.tsx:
```diff
@@ -264,22 +265,48 @@ export function panelMenuBehavior(menu: VizPanelMenu, isRepeat = false) {
+
+    items.push({
+      text: 'Extensions',
+      component: ExtensionsMenuItem,
```
I assume this will mean we need that scenes PR as well?

torkelo replied:
Instead of having to define a new component every time menu is opened I think it would be cleaner to have the component be a separate standalone component that takes in panel + dashboard as props. 

It would then be 
```tsx
   items.push({
      text: 'Extensions',
      component: <ExtensionsMenuItem panel={panel} dashboard={dashboard} />
```

maybe component should then be "reactNode", alternatively you have component + props. 

The thing that feels messy is having component and then having to create a new component (That captures the panel + dashboard variables via scoped lamda)

leventebalogh replied:
Thanks, I think that's a good idea! I'll try it out and get back!

leventebalogh commented:
/deploy-to-hg

leventebalogh commented:
/deploy-to-hg

leventebalogh commented:
/deploy-to-hg

sunker reviewed (commented):
Very nice @leventebalogh! Did some quick testing and things appears to be working as expected. Will do more thorough testing later.

Not necessarily for this PR, but I wonder if we should do more to validate the meta data in plugin.json now that we rely on it so heavily? For example there's no schema restriction on the id of the exposed component in the dependencies section.

leventebalogh commented on public/app/features/plugins/extensions/utils.tsx:
```diff
@@ -579,3 +580,90 @@ export const isExposedComponentMetaInfoMissing = (
+  Object.values(config.apps).filter((app) => pluginIds.includes(app.id));
+
+export const getAppPluginIdFromExposedComponentId = (exposedComponentId: string) => {
+  return exposedComponentId.replace(/^plugins\//, '').split('/')[0];
```
Hmm, that's true 🤔 Thanks!

leventebalogh replied:
Fixed in https://github.com/grafana/grafana/pull/86624/commits/59e35ba12d11311af418ff539a8d819d12cb7705.

leventebalogh commented on public/app/features/plugins/extensions/useLoadAppPlugins.tsx:
```diff
@@ -0,0 +1,31 @@
+    preloadPlugins(appConfigs).then(() => {
+      setIsLoading(false);
+    });
+  }, []); // eslint-disable-line react-hooks/exhaustive-deps
```
Yeah, I think that's a fair point, I'll add it 👍

leventebalogh replied:
I have updated the hook to use `useAsync()` in https://github.com/grafana/grafana/pull/86624/commits/67d0c8ed5f58bc47c738c790edf4bb9fc46badb0.

leventebalogh commented on pkg/services/featuremgmt/registry.go:
```diff
@@ -1602,6 +1602,13 @@ var (
+			Description:     "Enables lazy-loading app plugins (stops preloading app plugins that have extensions related metadata info in their plugin.json)",
+			Stage:           FeatureStageExperimental,
+			Owner:           grafanaPluginsPlatformSquad,
+			RequiresRestart: true,
```
Very nice catch, thanks mate!

leventebalogh replied:
Fixed in https://github.com/grafana/grafana/pull/86624/commits/d24f817653b1a4d6452981b8a0afac3f44d5d19e.

grafakus commented:
Hi @leventebalogh! Same question as @ifrost:

> Hi @leventebalogh 👋 This item has been on our board for some time (Tracing and Profiling). Is it work in progress? Do you need any input from Tracing & Profiling?

Any input you might want from us?

leventebalogh commented:
Hey @grafakus and @ifrost 👋 

Thanks for reaching out! Sorry that I have missed the previous comment. No, I think it's all good for now, but thanks! It's a bit weird, it looks like I cannot remove it from your board? 🤔

leventebalogh commented:
/deploy-to-hg

leventebalogh commented:
/deploy-to-hg

sunker commented on public/app/features/plugins/extensions/utils.tsx:
```diff
@@ -421,3 +425,221 @@ export function getLinkExtensionPathWithTracking(pluginId: string, path: string,
+  return Object.values(config.apps)
+    .filter((app) => app.preload)
+    .filter((app) => !awaitedPluginIds.includes(app.id))
+    .filter(
```
Usually don't care about these things, but since this happens at app bootstrap you may want to replace these three loops with a simple for loop to improve performance.

leventebalogh replied:
That's fair enough. I have updated it in https://github.com/grafana/grafana/pull/86624/commits/c131a5d59ae6c02e98ff51bbdb23509124c1d784 (instead of a for loop I kept using the `.filter()` method so I can "just return" with the results, but in this sense I think they are identical, they both loop through the list once.

sunker reviewed (commented):
Great stuff! Think some functions accidentally ended up in the wrong file after the merge conflict, but aside from that I think this one is good to go! 

To further protect ourselves from accidecanlly breaking the lazy loading of apps, we could have added an e2e test that verifies that the plugin's module.js file is _only_ loaded when necessary. WDYT?

leventebalogh commented:
> To further protect ourselves from accidecanlly breaking the lazy loading of apps, we could have added an e2e test that verifies that the plugin's module.js file is only loaded when necessary. WDYT?

That's a good point, I'll add it 👍

leventebalogh commented on e2e/test-plugins/grafana-extensionstest-app/tests/legacy/extensionPoints.getters.spec.ts:
```diff
@@ -4,6 +4,12 @@ import { ensureExtensionRegistryIsPopulated } from '../utils';
+test.use({
+  featureToggles: {
+    appPluginLazyLoading: false,
+  },
```
Yea, I was only adding it here to be "explicit", and to be more resilient to the change of the default value in the future - what do you think?

leventebalogh replied:
Sorry, I am actually being wrong. I have enabled it by default for E2E tests [here](https://github.com/grafana/grafana/blob/a50745ef54e487310a8fdec1939f1e87af42773a/playwright.config.ts#L24), and only disabling it for these getters tests.

With that being the case, we might be already testing both scenarios (lazy-loading and not lazy-loading)? I will check if we "tick" those conditions for lazy loading for at least one of the plugins and will add an extra test-case.

leventebalogh replied:
I am going to resolve this as I have removed the feature-toggle due to making the whole thing opt-out - plugins will only stop being preloaded once they remove "preload: true" from their plugin.json. Please undo the resolving if you have objections.

sunker replied:
Makes sense!

leventebalogh commented on public/app/features/plugins/extensions/utils.tsx:
```diff
@@ -421,3 +425,221 @@ export function getLinkExtensionPathWithTracking(pluginId: string, path: string,
+  }
+
+  return false;
+};
```
Phew, thanks for this catch. Fixed in https://github.com/grafana/grafana/pull/86624/commits/bc6ea60993c1d4763286b22ef421e1c148829bfa.

mckn commented on public/app/features/plugins/extensions/usePluginComponent.tsx:
```diff
@@ -16,11 +17,19 @@ export function usePluginComponent<Props extends object = {}>(id: string): UsePl
   const registry = useExposedComponentsRegistry();
   const registryState = useObservable(registry.asObservable());
   const pluginContext = usePluginContext();
+  const { isLoading: isLoadingAppPlugins } = useLoadAppPlugins(getExposedComponentPluginDependencies(id));
```
Q: do we need to handle an error case here as well?

leventebalogh replied:
Hmm, it's a good question! I was not doing it here because the [`preload()` function in `pluginPreloader.ts` ](https://github.com/grafana/grafana/blob/61ab73d72a077069c752d4811dea5525c66a05fc/public/app/features/plugins/pluginPreloader.ts#L38) is handling the errors, and as far as I understand there _shouldn't_ be any errors that make their way up there (in the general scenarios). I was kinda imagining this `useLoadAppPlugins()` hook like something that you can just call, and in your UI component you only need to care about if it's loading or not, everything else would be handled inside. What do you think?

mckn replied:
Yeah, just had a look at the underlying code and it looks good!

mckn reviewed (approved):
LGTM! Added one question but overall I think the code looks good. Have not tested it so let me know if I should take this for a spin before we merge it.

leventebalogh commented:
/deploy-to-hg

sunker commented on public/app/features/plugins/plugin_loader.ts:
```diff
@@ -166,21 +169,47 @@ export function importDataSourcePlugin(meta: DataSourcePluginMeta): Promise<Gene
-  const fallbackLoadingStrategy = meta.loadingStrategy ?? PluginLoadingStrategy.fetch;
-  return importPluginModule({
+export class DisabledAppPlugin extends Error {}
+export class NotAnAppPlugin extends Error {}
```
Are these two being used?

leventebalogh replied:
No, good point 👍 Removed in https://github.com/grafana/grafana/pull/86624/commits/4b4e36152b3ab43a1edee88703dd8ab050637de9.

sunker reviewed (approved):
Fantastic work on this @leventebalogh!

leventebalogh committed:
fix: remove unused types

leventebalogh committed:
fix: call `setComponentsFromLegacyExports()` after meta is initialised

jackw commented on public/app/features/plugins/tests/plugin_loader.test.ts:
```diff
@@ -55,14 +53,17 @@ describe('Load App', () => {
+    // Importing the same app again doesn't initialise it twice
+    const importedAppAgain = await importAppPlugin(meta);
+    expect(importedAppAgain).toBe(app);
+    expect(app.init).toHaveBeenCalledTimes(1);
```
Am I wrong or the previous test was asserting the `init` method was called twice?

leventebalogh replied:
Yeah, it's a tricky one. `app.calledTwice` was becoming true [in case the app plugin meta already got initialised](https://github.com/grafana/grafana/pull/86624/files#diff-d5eeb2b4949eb1aec48069a8cc8f6ab67d92013987fd83b3dfd8cdb16d96286fL24), and it was basically checking if the meta was already set before with the same value. Instead of that I was only checking here if the `init()` is not called again if the plugin was already imported (due to the cache), and that it returns the same app. What we could do is to also add a check for the `meta`?

```ts
expect(importedAppAgain.meta).toBe(meta);
```

jackw commented on public/app/features/plugins/pluginPreloader.ts:
```diff
@@ -15,67 +13,28 @@ export type PluginPreloadResult = {
-  startMeasure(eventName);
-  const promises = apps.filter((config) => config.preload).map((config) => preload(config));
-  const preloadedPlugins = await Promise.all(promises);
+const preloadedAppPlugins = new Set<string>();
+const isNotYetPreloaded = ({ id }: AppPluginConfig) => !preloadedAppPlugins.has(id);
+const markAsPreloaded = (apps: AppPluginConfig[]) => apps.forEach(({ id }) => preloadedAppPlugins.add(id));
```
Are these used to know what was already loaded? `markAsPreloaded` is called before the actual preloading? 

Also you could tap into the SystemJS registry to know if a plugin is already loaded. 🤔

leventebalogh replied:
> Are these used to know what was already loaded? markAsPreloaded is called before the actual preloading so I'm not sure.

Basically my intention was to introduce a "cache" here, so I don't run the preloading logic multiple times for the same plugin. I called it before the actual preloading to fight racing conditions (if I do it once all the promises are resolved, then there is a time window while these apps can be preloaded again). On the other I should have probably remove them from the list of preloaded plugins if their preloading fails?

> Also you could tap into the SystemJS registry to know if a plugin is already loaded. 🤔

Yeah, I think this relates somewhat to my intention above, I was basically trying to guard against entering the preloading logic for the same plugin multiple times, while the SystemJS registry is only populated once all the preloading is finished, which means that we would have a time window where the same plugin can be preloaded multiple times - I am not sure if that makes sense?

jackw replied:
Ok, sounds like it's probably best we get this merged as is and we circle back to it to figure out removing from the list if preloading fails / tapping into systemjs to guard against running preload logic for same plugin multiple times. 👍

jackw commented on public/app/features/plugins/plugin_loader.ts:
```diff
@@ -166,21 +169,44 @@ export function importDataSourcePlugin(meta: DataSourcePluginMeta): Promise<Gene
+  const { plugin = new AppPlugin() } = pluginExports;
+  plugin.setComponentsFromLegacyExports(pluginExports);
+  plugin.init(meta);
+  plugin.meta = meta;
```
And... you've already fixed it! 😄

jackw reviewed (approved):
Awesome work on this @leventebalogh 👏 

LGTM! 🚀

leventebalogh commented:
/deploy-to-hg
